src/HexagonOptimize.cpp

```diff
@@ -1605,6 +1616,293 @@ class OptimizeShuffles : public IRMutator {
+                    find_mpy_ops(op, Int(8, lanes), Int(8), 100, mpys, rest);
+                    vtmpy_suffix = ".vb.vb.b.b";
+                }
+            } else if (op->type.bits() == 32) {
+                find_mpy_ops(op, Int(16, lanes), Int(8), 100, mpys, rest);
+                vtmpy_suffix = ".vh.vh.b.b";
+            }
+
+            if (mpys.size() >= 3) {
+
+                int mpy_size = mpys.size();
+                vector<Expr> bases(mpy_size);
+                for(int i = 0; i < mpy_size; i++) {
+                    bases[i] = find_base(mpys[i].first);
+                }
+
+                // Checking all possible combinations for vtmpy reductions
+                for(int i = 0; i < mpy_size; i++) {
+
+                    if (!bases[i].defined() || !is_const(mpys[i].second, 1)) {
+                        continue;
+                    }
+                    Expr base2 = bases[i];
+                    for (int j = 0; j < mpy_size; j++) {
+                        Expr base1 = bases[j];
+                        if (!base1.defined() || !is_const(simplify(base2 - base1), 1)) {
+                            continue;
+                        }
+
+                        for (int k = 0; k < mpy_size; k++) {
+                            Expr base0 = bases[k];
+                            if (!base0.defined() || !is_const(simplify(base1 - base0), 1)) {
+                                continue;
+                            }
+
+                            Expr new_expr = native_interleave(Call::make(op->type, "halide.hexagon.vtmpy" + vtmpy_suffix, {mpys[k].first, mpys[i].first, mpys[k].second, mpys[j].second}, Call::PureExtern));
+                            Expr sum;
+                            for (int l = 0; l < mpy_size; l++) {
+                                if (l==i || l==j || l==k)
+                                    continue;
+                                if (sum.defined()) {
+                                    sum = sum + lossless_cast(op->type, mpys[l].first)*lossless_cast(op->type, mpys[l].second);
+                                } else {
+                                    sum = lossless_cast(op->type, mpys[l].first)*lossless_cast(op->type, mpys[l].second);
+                                }
+                            }
+                            if (sum.defined()) {
+                                new_expr = new_expr + sum;
+                            }
+                            if (rest.defined()) {
+                                new_expr = new_expr + rest;
+                            }
+                            expr = mutate(new_expr);
+                            return;
+                        }
+                    }
+                }
+            }
+        }
+
+        IRMutator::visit(op);
+    }
+
+public:
+    VtmpyGenerator(int native_vector_bytes) {
+        vector_size = native_vector_bytes;
+    }
+};
+
+
+class ModifyExpressions : public IRMutator {
```

Comment:
Please add a comment here that gives some details about what ModifyExpressions is trying to achieve because ModifyExpressions is a very generic name.